python/ml4ir/model/serving.py

```diff
@@ -0,0 +1,119 @@
+import tensorflow as tf
+from tensorflow import TensorSpec, TensorArray
+from ml4ir.config.keys import ServingSignatureKey
+from ml4ir.data.tfrecord_reader import make_parse_fn
+
+
+def define_default_signature(model, feature_config):
+    # Default signature
+    #
+    # TODO: Define input_signature
+    #
+    # @tf.function(input_signature=[])
+    # def _serve_default(**features):
+    #     features_dict = {k: tf.cast(v, tf.float32) for k, v in features.items()}
+    #     # Run the model to get predictions
+    #     predictions = model(inputs=features_dict)
+
+    #     # Mask the padded records
+    #     for key, value in predictions.items():
+    #         predictions[key] = tf.where(
+    #             tf.equal(features_dict['mask'], 0),
+    #             tf.constant(-np.inf),
+    #             predictions[key])
+
+    #     return predictions
+    return
+
+
+def define_tfrecord_signature(model, feature_config, pad_records):
+    """
+    Add signatures to the tf keras savedmodel
+
+    Returns:
+        Serving signature function that accepts a TFRecord string tensor and returns predictions
+    """
+
+    # TFRecord Signature
+    # Define a parsing function for tfrecord protos
+    inputs = feature_config.get_all_features(key="node_name", include_label=False)
+
+    """
+    NOTE:
+    Setting pad_records=False for tfrecord signature as it is used at inference time
+    and we do NOT want to score on padded records for performance reasons
+
+    Limitation: This limits the serving signature to only run inference on a single query
+    at a time given the current implementation. This is a tricky issue to fix because
+    there is no real way to generate a dense tensor of ranking scores from different queries,
+    as they might have varying number of records in each of them.
+
+    Workaround: To infer on multiple queries, run predict() on each of the queries separately.
+    """
+    tfrecord_parse_fn = make_parse_fn(
+        feature_config=feature_config,
+        max_num_records=25,
```

Comment:
we still have this "max" of 25?

Reply:
Yep. I was hoping you wouldn't catch this.

So, this path would only be invoked if you don't pass in a rank field in the data - then there is no way to guess the num_records in the query and it reverts back to this set `max_num_records`. Don't worry, it will not be invoked in our case since we have set `rank` as `required=True`. It will break earlier if you do not pass `rank` in the proto.